sql/src/main/java/io/crate/execution/engine/window/WindowProjector.java

```diff
@@ -83,23 +82,32 @@ public static WindowProjector fromProjection(WindowAggProjection projection,
                                                  Version indexVersionCreated,
                                                  IntSupplier numThreads,
                                                  Executor executor) {
-        LinkedHashMap<io.crate.expression.symbol.WindowFunction, List<Symbol>> functionsWithInputs = projection.functionsWithInputs();
-        ArrayList<WindowFunction> windowFunctions = new ArrayList<>(functionsWithInputs.size());
-        List<CollectExpression<Row, ?>> windowFuncArgsExpressions = new ArrayList<>(functionsWithInputs.size());
-        Input[][] windowFuncArgsInputs = new Input[functionsWithInputs.size()][];
-        int inputsIndex = 0;
-        for (Map.Entry<io.crate.expression.symbol.WindowFunction, List<Symbol>> functionAndInputsEntry : functionsWithInputs.entrySet()) {
+        var windowFunctionContexts = projection.windowFunctionContexts();
+        var numWindowFunctions = windowFunctionContexts.size();
+
+        ArrayList<WindowFunction> windowFunctions = new ArrayList<>(numWindowFunctions);
+        ArrayList<CollectExpression<Row, ?>> windowFuncArgsExpressions = new ArrayList<>(numWindowFunctions);
+        Input[][] windowFuncArgsInputs = new Input[numWindowFunctions][];
+
+        for (int idx = 0; idx < numWindowFunctions; idx++) {
+            var windowFunctionContext = windowFunctionContexts.get(idx);
+
             InputFactory.Context<CollectExpression<Row, ?>> ctx = inputFactory.ctxForInputColumns(txnCtx);
-            ctx.add(functionAndInputsEntry.getValue());
-            windowFuncArgsInputs[inputsIndex] = ctx.topLevelInputs().toArray(new Input[0]);
-            inputsIndex++;
-            windowFuncArgsExpressions.addAll(ctx.expressions());
+            ctx.add(windowFunctionContext.inputs());
 
-            FunctionImplementation impl = functions.getQualified(functionAndInputsEntry.getKey().info().ident());
+            FunctionImplementation impl = functions.getQualified(
+                windowFunctionContext.function().info().ident());
             if (impl instanceof AggregationFunction) {
+                Input<?> filter;
+                if (windowFunctionContext.filter().symbolType() == SymbolType.LITERAL) {
+                    filter = (Input<?>) windowFunctionContext.filter();
+                } else {
+                    ctx.add(windowFunctionContext.filter());
+                    filter = ctx.expressions().get(ctx.expressions().size() - 1);
```

Comment:
Is it always safe to assume that the last expression is the filter? I guess this depends on how the `InputFactory` context is traversing/processing the symbol?

Reply:
seems to be safe, the filter `InputColumn` is visited by InputColumnVisitor and added to the expression list by calling `ctx.add`. All other expressions are added to the list earlier. Any suggestion how it can be broken, i can't come up with a test for this scenario.